tests/unit/s2n_extended_master_secret_test.c

```diff
@@ -165,5 +171,139 @@ int main(int argc, char **argv)
         EXPECT_SUCCESS(s2n_config_free(config));
     }
 
+    /* Connection where the client supports EMS but the server does not support EMS */
+    {
+        struct s2n_config *config = s2n_config_new();
+        EXPECT_NOT_NULL(config);
+
+        /* TLS1.2 cipher preferences */
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "ELBSecurityPolicy-2016-08"));
+        EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(config));
+        struct s2n_cert_chain_and_key *chain_and_key = NULL;
+        EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key, S2N_DEFAULT_ECDSA_TEST_CERT_CHAIN,
+                                                    S2N_DEFAULT_ECDSA_TEST_PRIVATE_KEY));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+
+        struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT);
+        EXPECT_NOT_NULL(client_conn);
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+        struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER);
+        EXPECT_NOT_NULL(server_conn);
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+        /* Create nonblocking pipes */
+        struct s2n_test_io_pair io_pair = { 0 };
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+        /* Negotiate until server has read the Client Hello message */
+        s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+        EXPECT_OK(s2n_negotiate_until_message(client_conn, &blocked, SERVER_HELLO));
+        EXPECT_OK(s2n_negotiate_until_message(server_conn, &blocked, SERVER_HELLO));
+
+        /* s2n servers by default support EMS. We turn it off by manually setting ems_negotiated to false
+         * and removing the EMS extension from our received extensions. */
+        server_conn->ems_negotiated = false;
+        s2n_extension_type_id ems_ext_id = 0;
+        EXPECT_SUCCESS(s2n_extension_supported_iana_value_to_id(TLS_EXTENSION_EMS, &ems_ext_id));
+        S2N_CBIT_CLR(server_conn->extension_requests_received, ems_ext_id);
+
+        /* Connection is a successful and EMS is not negotiated */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_TRUE(IS_TLS12_CONNECTION(client_conn, server_conn));
+        EXPECT_FALSE(IS_EMS_NEGOTIATED(client_conn, server_conn));
```

Comment:
This will pass if either client or server did not negotiate EMS, but the other did. Is that sufficient, or do we need to confirm they both did not negotiate EMS?

Reply:
good point